discount-details-block/src/BlockExtension.liquid

```diff
@@ -0,0 +1,250 @@
+import { useState, useEffect } from 'react';
+// The target used here must match the target used in the extension's toml file (./shopify.extension.toml)
+const TARGET = 'admin.discount-details.function-metafields.render';
+
+export default reactExtension(TARGET, async (api) => {
+  return <App />;
+});
+
+function App() {
+  const { applyMetafieldChange, i18n, data } = useApi(TARGET);
+  const [percentage, setPercentage] = useState('');
+  const [error, setError] = useState('');
+  const initialMetafields = data?.metafields || [];
+  const [savedMetafields] = useState(initialMetafields);
+
+  useEffect(() => {
+    const transferPercentage = parsePercentageMetafield(
+      savedMetafields.find(
+        (metafield) => metafield.key === 'function-configuration',
+      )?.value,
+    );
+    setPercentage(transferPercentage);
+  }, [initialMetafields]);
+
+  // Handle percentage value change
+  const onPercentageValueChange = async (value) => {
+    setPercentage(value);
+    await applyExtensionMetafieldChange('percentage', value);
+  };
+  // Commit form values to metafields
+  async function applyExtensionMetafieldChange(key, value) {
+    const commitFormValues = {
+      percentage,
+      [key]: value,
+    };
+    const results = await applyMetafieldChange({
+      type: 'updateMetafield',
+      namespace: 'anyothernamespace',
+      key: 'function-configuration',
+      value: JSON.stringify(commitFormValues),
+      valueType: 'json',
+    });
+
+    if (results.status === 'error') {
+      setError(results.message);
+    }
+  }
+
+  return (
+    <>
+      <Form>
+        <FunctionSettings>
+          <Section>
+            <Box paddingBlockEnd="300">
+              <BlockStack gap="base">
+                <Text variant="headingMd" as="h2">
+                  {i18n.translate('percentage')}
+                </Text>
+                <NumberField
+                  label={i18n.translate('discountPercentage')}
+                  name="percentage"
+                  autoComplete="on"
+                  value={percentage}
+                  onChange={onPercentageValueChange}
+                  suffix="%"
+                />
+              </BlockStack>
+            </Box>
+          </Section>
+        </FunctionSettings>
+      </Form>
+    </>
+  );
+}
+async function makeGraphQLQuery(query, variables) {
+  const graphQLQuery = {
+    query,
+    variables,
+  };
+
+  const res = await fetch('shopify:admin/api/graphql.json', {
+    method: 'POST',
+    body: JSON.stringify(graphQLQuery),
+  });
+
+  if (!res.ok) {
+    console.error('Network error');
+  }
+
+  return await res.json();
+}
+
+function parsePercentageMetafield(value) {
+  try {
+    return JSON.parse(value).percentage;
+  } catch {
+    return 0;
+  }
+}
+
+{% else %}
+
+
+import {
+  FunctionSettings,
+  Text,
+  Form,
+  Banner,
+  NumberField,
+  Box,
+  BlockStack,
+  Section,
+  extension,
+} from '@shopify/ui-extensions/admin';
+
+// [START discounts-allocator.target]
```

Comment:
you won't want to have these comments in here, fyi 🙏 they're just for the tutorial